File: countPoints.py
# THE PROBLEM STATEMENT IS TO FIND THE NUMBER OF RODS THAT CONTAINS ALL THE THREE RINGS. THERE ARE N RINGS AND EACH RING IS EITHER RED, GREEN OR BLUE. THERE WILL BE A
# TOTAL OF TEN RODS NUMBERED FROM 0 TO 9. YOU ARE GIVEN A STRING OF LENGTN 2*N, EVERY TWO CHARCTERS IN THE STRING FORM A COLOR-POSITION PAIR THAT IS
# USED TO DESCRIBE THE POSITION OF THE RINGS.

# MY APPROACH IS TO USE DICTIONARY TO STORE THE WHICH RODS NUMBERED FROM 0 TO 9 CONTAINS ALL THE THREE RINGS AND THEN RETURN THE COUNT OF NUMBER OF RODS.

class Solution:
    def countPoints(self, rings: str) -> int:
        rods = {}
        ans = 0
        
        # INITIALISING THE DICTIONARY.
        for i in range(10):
            rods[i] = [0,0,0]
        
        # TRAVERSING THE STRING.
        for i in range(0,len(rings),2):
            rodColor = rings[i]
            rod = int(rings[i+1])
            
            if(rodColor == 'R'):
                rods[rod][0] = 1
            elif(rodColor == 'G'):
                rods[rod][1] = 1
            elif(rodColor == 'B'):
                rods[rod][2] = 1
            
        for i in range(10):
            if(sum(rods[i]) == 3):
                ans += 1
        
        return ans
      
# TIME COMPLEXITY : O(N)
# SPACE COMPLEXITY : O(1) SINCE THE NUMBER OF RODS DOEN'T GROW LINEARLY.

File: countPoints.java
// THE PROBLEM STATEMENT IS TO RETURN AN ARRAY IN WHICH AT EACH INDEX CONTAINS THE ANSWER OF THE EACH QUERY. I.E. LENGTH OF THE ARRAY TO BE RETURNED = LENGTH
// OF THE QUERY ARRAY. IN THE QUERY ARRAY, EACH INDEX CONTAINS THE CENTRE COORDINATES AND THE RADIUS OF A CIRCLE TO BE DRAWN AROUND THE CIRCLE.
// FOR EACH QUERY, WE HAVE TO FIND THE NUMBER OF POINTS IN THE PONTS ARRAY THAT IS WITHIN THE CIRCLE. ADDITIONALLY, FOR EACH QUERY WE HAVE TO TAKE ONLY O(N) TIME
// COMPLEXITY.

// MY APPROACH IS FOR EACH QUERY, TRAVERSE THE GIVEN POINTS ARRAY FIND THE DISTANCE OF CURRENT POINTS FROM THE CENTER AND APPLY THE PYTHAGORAS THEOREM. IF THE
// THIRD SIDE IS LESS THAN THE OR EQUAL TO THE RADIUS THEN INCREMENT THE PONITS COUNT BY ONE, AFTER GOING THROUGH ALL THE POINTS
// APPEND IT TO THE ARRAY. FINALLY, RETURN THE ARRAY.

class Solution 
{
    public int[] countPoints(int[][] points, int[][] queries) 
    {
        // DECLARING THE ARRAY TO BE RETURNED
        int[] ans = new int[queries.length];
        
        // TRAVERSING THE QUERIES ARRAY.
        for(int i=0;i<queries.length;i++)
        {
            int p = 0,r = queries[i][2];
            
            for(int j=0;j<points.length;j++)
            {
                int dx = Math.abs(queries[i][0]-points[j][0]),dy=Math.abs(queries[i][1]-points[j][1]);
                int c = (int)Math.ceil(Math.sqrt((dx*dx)+(dy*dy)));
                
                if(c <= r)
                {
                    p += 1;
                }
            }
            
            ans[i] = p;
        }
        
        return ans;
    }
}

// TIME COMPLEXITY : O(M*N) WHERE M IS THE LENGTH OF THE QUERIES ARRAY AND N IS THE LENGTH OF THE POINTS ARRAY.
// SPACE COMPLEXITY : O(M)
